File: packages/ruby/lib/opensms/resources/templates.rb
# frozen_string_literal: true

require_relative "base"

module Opensms
  module Resources
    # The +templates+ resource: reusable message bodies with {{placeholders}}.
    # Accessed as +client.templates+.
    class Templates < Base
      FIELDS = %i[name body traffic_type].freeze

      # GET /v1/templates -> Page<Template>.
      def list(params = nil, **kwargs)
        http_page("/v1/templates", Models.build(merge(params, kwargs), %i[limit cursor]))
      end

      # POST /v1/templates -> 201 Template.
      # @param params [Hash] :name (required), :body (required), :traffic_type
      def create(params = nil, idempotency_key: nil, **kwargs)
        body = Models.build(merge(params, kwargs), FIELDS, required: %i[name body])
        @transport.request(:post, "/v1/templates", body: body, idempotency_key: idem(idempotency_key))
      end

      # GET /v1/templates/{id} -> Template.
      def get(id)
        http_get(path("/v1/templates", id!(id)))
      end

      # PATCH /v1/templates/{id} -> Template.
      def update(id, params = nil, **kwargs)
        body = Models.build(merge(params, kwargs), FIELDS)
        @transport.request(:patch, path("/v1/templates", id!(id)), body: body)
      end

      # DELETE /v1/templates/{id} -> 204.
      # @return [nil]
      def delete(id)
        http_delete(path("/v1/templates", id!(id)))
      end
    end
  end
end

File: packages/python/src/opensms/resources/templates.py
"""The ``templates`` resource."""

from __future__ import annotations

from typing import Optional

from .._serialize import path_id, prune
from ..models import Page, Template
from ._base import Resource


class Templates(Resource):
    """Reusable message templates with ``{{name}}`` placeholders. Accessed as
    ``client.templates``."""

    def list(self, *, limit: Optional[int] = None, cursor: Optional[str] = None) -> Page[Template]:
        """List templates."""
        return self._page(self._t.request("GET", "/v1/templates", query={"limit": limit, "cursor": cursor}))

    def create(
        self,
        *,
        name: str,
        body: str,
        traffic_type: Optional[str] = None,
        idempotency_key: Optional[str] = None,
    ) -> Template:
        """Create a template. ``variables`` in the result lists its placeholders."""
        payload = prune({"name": name, "body": body, "traffic_type": traffic_type})
        return self._t.request(
            "POST", "/v1/templates", json_body=payload, idempotent=True, idempotency_key=idempotency_key
        )

    def get(self, id: str) -> Template:
        """Fetch a template."""
        return self._t.request("GET", f"/v1/templates/{path_id(id)}")

    def update(
        self,
        id: str,
        *,
        name: Optional[str] = None,
        body: Optional[str] = None,
        traffic_type: Optional[str] = None,
    ) -> Template:
        """Update the given fields (PATCH)."""
        payload = prune({"name": name, "body": body, "traffic_type": traffic_type})
        return self._t.request("PATCH", f"/v1/templates/{path_id(id)}", json_body=payload)

    def delete(self, id: str) -> None:
        """Delete a template."""
        self._t.request("DELETE", f"/v1/templates/{path_id(id)}")
